### DISCUSSION TITLE
Missing auth from OpenAPI Schema

### DISCUSSION THREAD

**Discussion by jerinpetergeorge** [encode/django-rest-framework#7929] [Question & Answer]
I have set up swagger doc using our documentation ([Schema-(Doc)][1] and [Documenting your API-(Doc)][2]) and I've got the result as below
![image](https://user-images.githubusercontent.com/21334929/115101543-b9568a80-9f62-11eb-861b-477330778ba5.png)

Looks good!!!

But, the generated UI doesn't have any [security options(as per the Open API schema)][3]. Did I miss something here? :thinking: 

Ideally, looking for this ***Authorize*** button :point_down: :point_down: 
![image](https://user-images.githubusercontent.com/21334929/115101748-5cf46a80-9f64-11eb-8e85-f1c33df5ed48.png)

[1]:https://www.django-rest-framework.org/api-guide/schemas/
[2]:https://www.django-rest-framework.org/topics/documenting-your-api/
[3]:https://swagger.io/docs/specification/authentication/

**Comment by jerinpetergeorge** [comment]
I'm not sure whether this functionality is available in the native DRF doc or not. I hope you guys can be more certain in this @tomchristie @carltongibson 

(sorry for the noise :) )

**Reply by carltongibson** [reply]
Hi @jerinpetergeorge — good question! 

This just isn't implemented yet. Something to do with the limits of human power vs the wide scope of the OpenAPI spec. 

It would be a nice addition though. 

In the mean time, best to subclass `SchemaGenerator` to add the security options.

**Reply by n2ygk** [reply]
See #7516

**Comment by n2ygk** [comment]
See https://github.com/encode/django-rest-framework/pull/7516

**Reply by n2ygk** [reply]
I contributed #7516 about 2 years ago. Still pending 3.14 milestone @carltongibson. I'm pretty sure this begins to address @jerinpetergeorge's request, along with pending work to extend this to Django-oauth-toolkit. It's why I implemented it, also using swagger-ui.

**Comment by monkut** [comment]
Until something gets rolled in here's a hard-coded solution I used in my case where my API uses JWT token authentication.

I created a custom renderer based off of the `OpenAPIRenderer` to force in the related security scheme.

```python
from rest_framework.compat import yaml
from rest_framework.renderers import BaseRenderer


class OpenAPIWithJWTBearerSecuritySchemeRenderer(BaseRenderer):
    media_type = 'application/vnd.oai.openapi'
    charset = None
    format = 'openapi'

    def __init__(self):
        assert yaml, 'Using OpenAPIRenderer, but `pyyaml` is not installed.'

    def render(self, data, media_type=None, renderer_context=None):
        # disable yaml advanced feature 'alias' for clean, portable, and readable output
        class Dumper(yaml.Dumper):
            def ignore_aliases(self, data):
                return True
        data["security"] = [{"bearerAuth": []}]
        assert "components" in data      
        data["components"]["securitySchemes"] = {"bearerAuth": {"type": "http", "scheme": "bearer", "bearerFormat": "JWT"}}
        return yaml.dump(data, default_flow_style=False, sort_keys=False, Dumper=Dumper).encode('utf-8')
```
